src/Adapter/Product/Pack/QueryHandler/GetPackedProductsHandler.php

```diff
@@ -28,44 +28,115 @@
+     * @param ProductImageRepository $productImageRepository
      */
-    public function __construct(int $defaultLangId)
-    {
-        $this->defaultLangId = $defaultLangId;
+    public function __construct(
+        int $defaultLangId,
+        ProductPackRepository $productPackRepository,
+        AttributeRepository $attributeRepository,
+        CombinationNameBuilder $combinationNameBuilder,
+        ProductImageRepository $productImageRepository
+    ) {
+        $this->productRepository = $productPackRepository;
+        $this->combinationNameBuilder = $combinationNameBuilder;
+        $this->attributeRepository = $attributeRepository;
+        $this->languageId = $defaultLangId;
+        $this->productImageRepository = $productImageRepository;
     }
 
     /**
      * {@inheritdoc}
      */
     public function handle(GetPackedProducts $query): array
     {
-        $packId = $query->getPackId()->getValue();
+        $packedItems = $this->productRepository->getPackedProducts(
+            $query->getPackId(),
+            $query->getLanguageId()
+        );
+        $packedProducts = [];
+        $combinationIds = [];
+        foreach ($packedItems as $packedItem) {
+            $combinationId = (int) $packedItem['id_product_attribute_item'];
 
-        $packedItems = Pack::getItems($packId, $this->defaultLangId);
+            if ($combinationId > 0) {
+                $combinationIds[] = new CombinationId($combinationId);
+            }
+        }
+        $attributesInformations = $this->attributeRepository->getAttributesInfoByCombinationIds(
+            $combinationIds,
+            new LanguageId($this->languageId)
+        );
 
-        $packedProducts = [];
         foreach ($packedItems as $packedItem) {
-            $packedProducts[] = new PackedProduct(
-                (int) $packedItem->id,
-                (int) $packedItem->pack_quantity,
-                (int) $packedItem->id_pack_product_attribute
+            $combinationId = (int) $packedItem['id_product_attribute_item'];
+            if ($combinationId === NoCombinationId::NO_COMBINATION_ID) {
+                $coverUrl = $this->productImageRepository->getProductCoverUrl(
+                    new ProductId((int) $packedItem['id_product_item'])
+                );
+            } else {
+                $coverUrl = $this->productImageRepository->getCombinationCoverUrl(
+                    new CombinationId($combinationId)
+                );
+            }
+            $name = $packedItem['name'];
+            if ($combinationId > 0) {
+                $name = $this->combinationNameBuilder->buildFullName($name, $attributesInformations[$combinationId]);
+            }
+
+            $packedProducts[] = new PackedProductDetails(
+                (int) $packedItem['id_product_item'],
+                (int) $packedItem['quantity'],
+                (int) $combinationId,
+                (string) $name,
+                (string) $packedItem['reference'],
```

Comment:
I don't think you should return the reference independently Instead the reference should be handled and added by the `CombinationNameBuilder->buildFullName` method, so that it's part of the name

This way the reference will be visible in the form data on refresh, but also in the API results And the content will be fully replaced when you select on an element from the entity search

Because I didn't test the PR but I would say right now, the suggestion have no references, when you select one you still have no reference but once you submit the reference appears in the rendered packs right?